worker-workflow/src/main/java/com/github/cafdataprocessing/WorkflowWorker.java

```
@@ -0,0 +1,182 @@
+    @Override
+    public void processDocument(final Document document) throws InterruptedException, DocumentWorkerTransientException
+    {
+        // Get the worker task properties passed in custom data
+        final ExtractedProperties extractedProperties = CustomDataExtractor.extractPropertiesFromCustomData(document);
+        if(!extractedProperties.isValid()) {
+            LOG.warn("Custom data on document is not valid for this worker. Processing of this document will not " +
+                    "proceed for this worker.");
+            return;
+        }
+        final String projectId = extractedProperties.getProjectId();
+        final long workflowId = extractedProperties.getWorkflowId();
+
+        // Get the specified workflow and transform it to a JavaScript representation
+        TransformWorkflowResult transformWorkflowResult = workflowCache.getTransformWorkflowResult(workflowId, projectId);
+        if(transformWorkflowResult==null) {
+            ReentrantLock transformWorkflowLock = workflowCache.getTransformWorkflowLock(workflowId, projectId);
+            transformWorkflowLock.lock();
+            try {
+                transformWorkflowResult = workflowCache.getTransformWorkflowResult(workflowId, projectId);
+                if (transformWorkflowResult == null) {
+                    transformWorkflowResult = transformWorkflow(document, extractedProperties);
+                    if (transformWorkflowResult == null) {
+                        LOG.warn("Failure during workflow transformation. Processing of this document will not proceed " +
+                                "for this worker.");
+                        return;
+                    }
+                    // Cache the transform result
+                    workflowCache.addTransformWorkflowResult(workflowId, projectId, transformWorkflowResult);
+                }
+            }
+            finally {
+                transformWorkflowLock.unlock();
+            }
+        }
```

The logic here seems sound (it doesn't suffer Java's usual double-locking pattern issue - see [Double-checked locking: Clever, but broken](https://www.javaworld.com/article/2074979/java-concurrency/double-checked-locking--clever--but-broken.html) if you don't know what I'm talking about), but it's also overly complicated - you are using the Guava Cache object but you would get the extra functionality you're hand-coding with the `ConcurrentHashMap` if you used the Guava LoadingCache instead.  See [LoadingCache.html#get(K)](https://google.github.io/guava/releases/19.0/api/docs/com/google/common/cache/LoadingCache.html#get(K)).

---

I had started with LoadingCache but couldn't think of a way to add  specific failures during transformation to a document if the loader was in charge of performing the transformation (it would only have the key information available to it). Will take another look at it.

---

The answer to that is with an exception.  If you can't load the cache element then you can throw an exception which you can catch back in the code where you have the document context.